Classes/Wwwision/Neos/MailChimp/Domain/Service/MailChimpService.php

```diff
@@ -185,32 +206,32 @@ private function patch($resource, array $arguments = null)
      * @param string $resource The REST resource name (e.g. "lists")
      * @param array|null $arguments Arguments to be send to the API endpoint
      * @return array The decoded response
-     * @throws MailChimpException
+     * @throws ResourceNotFoundException | MailChimpException | HttpException
      */
-    private function makeRequest($method, $resource, array $arguments = null)
+    private function makeRequest(string $method, string $resource, array $arguments = null): array
     {
-        $uri = new Uri($this->apiEndpoint . '/' . $resource);
+        $uri = $this->uriFactory->createUri($this->apiEndpoint . '/' . $resource);
         if ($method === 'GET' && $arguments !== null) {
-            $uri->setQuery(http_build_query($arguments));
+            $uri = $uri->withQuery(http_build_query($arguments));
         }
-        $request = Request::create($uri, $method);
-        $request->setHeader('Accept', 'application/vnd.api+json');
-        $request->setHeader('Content-Type', 'application/vnd.api+json');
-        $request->setHeader('Authorization', 'apikey ' . $this->apiKey);
+        $request = $this->serverRequestFactory->createServerRequest($method, $uri)
+            ->withHeader('Accept', 'application/vnd.api+json')
+            ->withHeader('Content-Type', 'application/vnd.api+json')
+            ->withHeader('Authorization', 'apikey ' . $this->apiKey);
         if ($method !== 'GET' && $arguments !== null) {
-            $request->setContent(json_encode($arguments));
+            $request = $request->withParsedBody($arguments);
```

Comment:
This does not work… while it sets the parsed body, the `CurlEngine` when used to submit the request uses `getBody()->getContent()` to populate `CURLOPT_POSTFIELDS`…

Reply:
OK, thanks! I already expected this to be the bug but couldn't test yet. I'll take care on Monday if nobody else does before

Reply:
Oh, slightly different, but still: Since we use PUT here, this kicks in in `CurlEngine`:

```php
        // If the content is a stream resource, use cURL's INFILE feature to stream it
          $content = $request->getBody()->getContents();
  
          // some more code here for other methods
            case 'PUT':
                curl_setopt($curlHandle, CURLOPT_PUT, true);
                if ($content !== '') {
                    $inFileHandler = fopen('php://temp', 'r+');
                    fwrite($inFileHandler, $content);
                    rewind($inFileHandler);
                    curl_setopt_array($curlHandle, [
                        CURLOPT_INFILE => $inFileHandler,
                        CURLOPT_INFILESIZE => strlen($content),
                    ]);
                }
```

Reply:
Well, tried to use `withBody` and a `ContentStream`, but simply creating that from a string is so convoluted… 😱 I give up for today.

Reply:
Yes, it feels a bit cumbersome.. I'll look into it